requests-for-comments/2022-03-08-full-saml.md

```diff
@@ -0,0 +1,48 @@
+
+
+## Proposal
+
+1. SAML configuration becomes organization-based instead of instance-based. We create a separate model for `OrganizationConfiguration` where we track all SAML parameters (ACS URL, Entity ID, X509 certificate and attribute names [4]), and of course drop all configuration from instance environment variables.
+2. For the next release, when updating, if SAML configuration is set and the instance has a single organization, we set the same SAML parameters for that organization.
+3. Drop `SAML_ENFORCED` in favor of a `SSO_ENFORCEMENT` attribute that supports `saml`, `google-oauth2`, `github` & `gitlab` to solve [#7535](https://github.com/PostHog/posthog/issues/7535) too (only one per organization).
+4. Drop `SAML_DISABLED` support to make maintenance easier and avoid having multiple build flows. Drawback is that the FOSS version which does not support SAML will still require the SAML dependencies.
+5. If SAML is enabled for at least one organization in the instance, the login page shows only the email input at first. Upon entering an email, we do an API request as shown below (where `enforced_sso` contains information on whether to enforce log in with a particular SSO provider [either SAML, Google, ...] or `null` if the user has no SSO enforced **or the user does not exist**).
+    ```
+    GET /api/users/pre-login
+    {
+        "email": "test@posthog.com"
+    }
+
+    Response:
+    {
+        "enforced_sso": "saml"
+    }
+    ```
+
+    > ⚠️ Security concern. If the user has any SSO enforced, it will leak information about a specific user being registered in the PostHog instance.
+6. If the user belongs only to an organization with SSO enforced (or multiple organizations with the same SSO enforcement policy), we won't allow logging in with password, changing password or log in with any other SSO mechanism. If the user requests a password reset we'll only include the information about which SSO provider to use but no link to reset the password.
+7. Bugs [#5759 Respect redirect URL when logging in with SAML](https://github.com/PostHog/posthog/issues/5759) & [#5839 Fix team member invites with SAML](https://github.com/PostHog/posthog/issues/5839) should be addressed too.
+8. Generally, users with SSO enforced will likely only belong to a single organization (because you're probably using your work email). However, there's the edge case of a user belonging to multiple organizations with different SSO policies. Here's how we'll address them.
+    - If the user attempts to join an organization that has a different SSO policy from the one present in their current organization(s), we won't allow the user to join the new organization and show a warning explaining why.
```

Comment:
If I understand well, the issue we're trying to avoid is somebody malicious logging in as `bob@posthog.com` with a self-hosted or compromised SAML server... that will make us trust that they're actually who they say they are, and hence give them access to other orgs.

- This might be technically trickier, but what about showing another login prompt then? The goal being that you can log in to each org individually. Or should we log you out of one org before logging in to another?
- Should we also show e.g. a lock next to the organization in the org dropdown?
- Should we even restrict domains, so that only users with `@posthog.com` domains allowed for org X, and no other user anywhere in the app can use this domain? (edit: saw the last line in the RFC... so that's a "not yet")